README
This is a MATE panel applet which allows you to enjoy football live scores
feed. 


USAGE
=====

You can select one or more leagues to check. The applet will notify you on each
of the following events: commenicing of a game, half time, game's end and goal
scored. To select which leagues to follow, right-click the applet.
Notifications will be shown up to 3 a time. 

In addition, at  start-up, the applet will notify you about any game in 
progress from the leagues you have selected. 

You can also view all the currents scores. To do so, left-click the applet.
The window will refresh each time an update is received. If you expand one or 
more leagues, this will be rememered between window closures and also between 
applet restarts. When applet is restarted, the previously expanded leagues 
will be on top of the list. 

When a game is in progress, a coloured dot will let you quickly check its
status: green means first half, yellow means second half before the 80th 
minute and red means after 80th minute (and extra time).

Past games will be stored in the list for 16 hours after their original start
time. Future games will appear on the list up to 16 fours before they begin.

The next tab of the same window lists the last 30 scored goals in all matches.


DATA SOURCES
============

Real time sports scores are big business, so nobody is willing to give out his
data. Ideally, the applet should use a source of data in machine-readable
format (JSON, RSS, XML...). Unfortunately, I was not able to find any service
which will provide such data free of charge. The applet now has to parse web
pages, which inevitably makes it susceptible to errors on every change of the
page layout. If you know of any good source of machine-readable live score 
that is available for free (even in the expense of viewing ads), please, let
me know (see AUTHORS for contact info).


HACKING
=======

Yes, you can write your very own data (feed) provider. You need to compile it
as shared library and put it in LIBDIR/mate-applet-livescore directory. It 
should be named lib_feed_XYZ.so where XYZ is whatever string you like.

The entry point of your library should be named "feed_main" and should look
like this:

	int feed_main(match_data **, int *);

The first argument is a double pointer to already malloc()'ed structure of
type "match_data" which size is sizeof(match_data). If you have N matches, 
you need to realloc() it to make it N times bigger; then put your matches 
inside as an array. 

The function should dereference the second pointer and set it to the number 
matches (i.e. the size of the array that the first pointer points to). 
Initially it will have a value of 0.

Function "feed_main" should return 1 on success or 0 on failure. 

The "match_data" structure has the following members (see src/applet.h):
-- char team_home[64] - the name of the home team, UTF-8
-- char team_away[64] - the name of the away team, UTF-8
-- int score_home - number of goals for the home team; will be ignored for 
   matches that have not yet commenced (set to 0 for them)
-- int score_away - same as previous, for away team
-- time_t start_time - match start time, UNIX timestamp, UTC (or current time)
   if actual start time is unknown.
-- int match_time - number of minutes since the beginning of match; will be 
   ignored for matches that have not yet commenced (set to 0 for them); when
   additional time is given in the end of first ahd second half, keep this
   value at 45 or 90. 
-- int match_time_added - number of minutes since the added time in the end of
   first or second half has began; will be ignored for matches that have not 
   yet commenced (set to 0 for them); reset to 0 when the first half or the 
   match ends.
-- int status - current match status, must be on of these: MATCH_NOT_COMMENCED,
   MATCH_FIRST_TIME, MATCH_HALF_TIME, MATCH_SECOND_TIME, MATCH_EXTRA_TIME,
   MATCH_FULL_TIME
-- int league_id - value will be ignored; set it to 0.
-- char league_name[256] - value will be ignore;, set it to '\0'
-- gboolean used - value will be ignored; set it to TRUE.

If you need to fetch something over HTTP, you can use the get_url() function
from http.c. It will get the URL for you and place it in the file you
specified. Set USER_AGENT to NULL to use the system default string. 

	int get_url (char *URL, char *USER_AGENT, char *FILENAME);

You can also make use of the trimming functions to remove spaces or quotes 
before and after a string:

	char *trim(char *s);
	char *trim_quotes(char *s);



IMPORTANT! This free software is released with hope that it will be useful, 
but with no warranty of any type at all. Use it at your own risk. You are free
to distribute it as well as to modify it as long as you preserve the original 
License. See LICENSE file for details.




ChangeLog
Jun 13, 2017 - Version 0.4.1
============================
* Fixes in second version of Ombibet feed

Apr 17, 2017 - Version 0.4.0
============================
* Added second version of Ombibet feed

Aug 2, 2016 - Version 0.3.1
============================
* Improve Extra Time and interrupted games handling in default provider.
* Handle the rare case when a reported call gets cancelled. 

Feb 24, 2015 - Version 0.3.0
============================
* Added Omnibet as feed provider and set it as default.

Jul 29, 2014 - Verzion 0.2.1
============================
* Show minute when the goal is scored in notification.
* Handle 'no information yet' status for matches which should have began.
* Fix blank notifications for teams which name contains an ampersand.
* Fix matches window to not periodically collapse opened leagues.

Jul 11, 2014 - Version 0.2.0
===========================
* Added InetPulse as second feed provider and set it as default.

May 17, 2014 - Version 0.1.3
============================
* Adapted IDDAA feed to change it time format for additional time.

Jan 1, 2014 - Version 0.1.2
===========================
* Fixed match duplication in main window

Dec 26, 2013 - Version 0.1.1
============================
* Fixed match name truncation in notifications.

Nov 25, 2013 - Version 0.1.0
============================
* Added info on goals for each match (info is only available on goals, scored
after the applet started).
* Added second tab to GUI - shows last 30 goals.
* Added notifications upon start-up about ongoing matches.
* Added handling of multiple goals in a match between two updates.

Nov 3, 2013 - Version 0.0.2
===========================
* Backported to GNOME2 (RHEL 6 and derivatives like CentOS, SL...).

Nov 1, 2013 - Version 0.0.1
============================
* Initial release.



src/applet.h
/*
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Library General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301,
 *  USA.
 *
 *  MATE Livescore applet written by Assen Totin <assen.totin@gmail.com>
 *  
 */

#include <string.h>
#include <unistd.h>
#include <libintl.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <glib.h>
#include <stdlib.h>
#include <wait.h>
#include <unistd.h>
#include <pwd.h>
#include <fcntl.h>
#include <errno.h>
#include <time.h>
#include <libxml/HTMLparser.h>
#include <libxml/HTMLtree.h>
#include <libsoup/soup.h>
#include <dlfcn.h>
#include <ctype.h>

#ifdef HAVE_MATE
	#include <mate-panel-applet.h>
#elif HAVE_GNOME_2
	#include <panel-applet.h>
	#include <panel-applet-gconf.h>
#endif

#ifdef HAVE_LIBMATENOTIFY
	#include <libmatenotify/notify.h>
#elif HAVE_LIBNOTIFY
	#include <libnotify/notify.h>
#endif

#ifdef HAVE_MATE
	#define APPLET_FACTORY "LivescoreAppletFactory"
	#define APPLET_ID "LivescoreApplet"
#elif HAVE_GNOME_2
	#define APPLET_FACTORY "OAFIID:LivescoreApplet_Factory"
	#define APPLET_ID "OAFIID:LivescoreApplet"
#endif


#define _(String) gettext (String)
#define APPLET_NAME "livescore"
#define APPLET_ICON_STATIC "applet_livescore_icon.png"
#define APPLET_IMAGE_RED "applet_livescore_red.png"
#define APPLET_IMAGE_YELLOW "applet_livescore_yellow.png"
#define APPLET_IMAGE_GREEN "applet_livescore_green.png"
#define APPLET_IMAGE_GRAY "applet_livescore_gray.png"
#define APPLET_IMAGE_NOTIF_GOAL "applet_livescore_goal.png"
#define APPLET_IMAGE_NOTIF_WHISTLE "applet_livescore_whistle.png"
#define APPLET_IMAGE_NOTIF_T1 "applet_livescore_t1.png"
#define APPLET_IMAGE_NOTIF_T2 "applet_livescore_t2.png"
#define APPLET_WINDOW_MATCHES_WIDTH 640
#define APPLET_WINDOW_MATCHES_HEIGHT 480
#define APPLET_WINDOW_SETTINGS_WIDTH 480
#define APPLET_WINDOW_SETTINGS_HEIGHT 320
#define APPLET_KEEP_TIME_MATCH 57600	// 16 hours
#define APPLET_SHOW_LAST_GOALS 30
#define APPLET_FEED_DEFAULT "lib_feed_omnibet.so"
// GSettings
#define APPLET_GSETTINGS_SCHEMA "org.mate.panel.applet.LivescoreApplet"
#define APPLET_GSETTINGS_PATH "/org/mate/panel/objects/livescore/"
#define APPLET_GSETTINGS_SEPARATOR "|"
#define APPLET_GSETTINGS_KEY_FAV "leagues-notify"
#define APPLET_GSETTINGS_KEY_EXP "leagues-expanded"
#define APPLET_GSETTINGS_KEY_FEED "feed"

// Menu strings
#ifdef HAVE_MATE
static const gchar *ui = 
"<menuitem name='MenuItem1' action='Settings' />"
"<menuitem name='MenuItem2' action='About' />"
;
#elif HAVE_GNOME_2
static const gchar *ui =
"<popup name='button3'>"
"<menuitem name='MenuItem1' verb='Settings' label='Settings'/>"
"<menuitem name='MenuItem2' verb='About' label='About' pixtype='stock' pixname='gnome-stock-about'/>"
"</popup>"
;
#endif

#ifdef HAVE_MATE
        typedef MatePanelApplet MyPanelApplet;
	typedef MatePanelAppletBackgroundType MyPanelAppletBackgroundType;
#elif HAVE_GNOME_2
        typedef PanelApplet MyPanelApplet;
	typedef PanelAppletBackgroundType MyPanelAppletBackgroundType;
#endif

// Two hidden colums will help manage the font weight for some cells
enum {
	COL_PIC = 0,
	COL_TIME,
	COL_SCORE,
	COL_MATCH,
	COL_HIDDEN_BOLD,
	COL_HIDDEN_BOOLEAN,
	NUM_COLS
};

enum {
	COL_GOALS_PIC = 0,
	COL_GOALS_TIME,
	COL_GOALS_SCORE,
	COL_GOALS_MATCH,
	NUM_COLS_GOAL
};

enum {
	MATCH_NOT_COMMENCED = 0,
	MATCH_FIRST_TIME,
	MATCH_HALF_TIME,
	MATCH_SECOND_TIME,
	MATCH_EXTRA_TIME,
	MATCH_FULL_TIME
};

enum {
	NOTIF_SHOW_IMAGE_WHISTLE = 0,
	NOTIF_SHOW_IMAGE_GOAL,
	NOTIF_SHOW_IMAGE_T1,
	NOTIF_SHOW_IMAGE_T2
};

typedef struct f_data {
	void *node_data;
	struct f_data *node_next;
} fifo_data;

typedef struct {
	fifo_data *fifo_head;
	fifo_data *fifo_tail;
} fifo;

typedef struct {
	char title[256];
	char body[256];
	int image;
	gboolean block;
} notif_data;

typedef struct {
	int goal_id;
	int match_id;
	int score_home;
	int score_away;
	int match_time;
	int match_time_added;
	time_t time_added;
	gboolean used;
	gboolean scored_home;
} goal_data;

typedef struct {
	int feed_id;
	char feed_name[256];
	gboolean enabled;
	gboolean selected;
} feed_data;

typedef struct {
	int league_id;
	char league_name[256];
	gboolean used;
	gboolean favourite;
	gboolean expanded;
} league_data;

typedef struct {
	int league_id;
	int match_id;
	char league_name[256];
	char team_home[64];
	char team_away[64];
	int score_home;
	int score_away;
	int status;
	time_t start_time;
	int match_time;
	int match_time_added;
	gboolean used;
} match_data;

typedef struct {
	GMainLoop *loop;
	MyPanelApplet *applet;
	GtkActionGroup *action_group;
	GtkWidget *image;
	GtkWidget *event_box;
	match_data *all_matches;
	league_data *all_leagues;
	feed_data *all_feeds;
	goal_data *all_goals;
	int all_matches_counter;
	int all_leagues_counter;
	int all_feeds_counter;
	int all_goals_counter;
	void *feed_handle;
	void (*feed_main)(match_data **, int *);
	gboolean dialog_matches_is_visible;
	fifo *notif_queue;
#ifdef HAVE_MATE
	GSettings *gsettings;
#endif
	GtkTreeStore *tree_store;
	GtkWidget *tree_view;
	GtkTreeStore *tree_store_goals;
	GtkWidget *tree_view_goals;
	GtkWidget *dialog_matches;
	GtkWidget *dialog_settings;
	GdkPixbuf *running_image_red;
	GdkPixbuf *running_image_green;
	GdkPixbuf *running_image_yellow;
	GdkPixbuf *running_image_gray;
	GdkPixbuf *notif_image_goal;
	GdkPixbuf *notif_image_whistle;
	GdkPixbuf *notif_image_t1;
	GdkPixbuf *notif_image_t2;
} livescore_applet;

// util.c
void show_notification (gchar *, gchar *, GdkPixbuf *);
gboolean cp(const char *, const char *);
char *trim(char *);
char *trim_quotes(char *);
char *first_word(char *);
char *string_ends(char *, int);
void debug(char *);
void debug_i(int);
void debug_u(unsigned int);

//menu.c
void quitDialogClose(GtkWidget *, gpointer);
void menu_cb_settings(GtkAction *, livescore_applet *);
void menu_cb_about(GtkAction *, livescore_applet *);

// gui.c
void gui_create_view_and_model(livescore_applet *);
void gui_create_view_and_model_goals(livescore_applet *);
void gui_update_model(livescore_applet * applet);
void gui_update_model_goals(livescore_applet * applet);
gboolean on_left_click (GtkWidget *, GdkEventButton *, livescore_applet *);

// manager.c
//gboolean manager_main(livescore_applet *, match_data *);
int manager_timer(livescore_applet *);
gboolean manager_populate_feed(livescore_applet *, gchar *, gboolean);

// http.c
int get_url (char *, char *, char *, GSList *, GSList **);

// fifo.c
fifo *fifo_new(void);
void fifo_add(fifo *, void *);
void *fifo_remove(fifo *);
void fifo_free(fifo *);
gboolean fifo_is_empty(fifo *);

// feed_iddaa.c
int feed_iddaa_main(livescore_applet *);

// feed_enetpulse.c
int feed_enetpulse_main(livescore_applet *);

// feed_omnibet.c
int feed_omnibet_main(livescore_applet *);

// main.c
#ifdef HAVE_GTK2
void applet_back_change (MyPanelApplet *, MyPanelAppletBackgroundType, GdkColor *, GdkPixmap *, livescore_applet *);
#elif HAVE_GTK3
void applet_back_change (MyPanelApplet *, MyPanelAppletBackgroundType, GdkRGBA *, cairo_pattern_t *, livescore_applet *);
#endif


void applet_destroy(MyPanelApplet *, livescore_applet *);

// Menu skeleton
#ifdef HAVE_MATE
static const GtkActionEntry applet_menu_actions_mate[] = {
	{ "Settings", GTK_STOCK_EXECUTE, "_Settings", NULL, NULL, G_CALLBACK (menu_cb_settings) },
	{ "About", GTK_STOCK_ABOUT, "_About", NULL, NULL, G_CALLBACK (menu_cb_about) }
};
#elif HAVE_GNOME_2
static const BonoboUIVerb applet_menu_actions_gnome [] = {
        BONOBO_UI_UNSAFE_VERB ("Settings", menu_cb_settings),
        BONOBO_UI_UNSAFE_VERB ("About", menu_cb_about),
        BONOBO_UI_VERB_END
};
#endif
